FAPI 0.86.1 sync

rinOfTheStars opened the pull request:
Is _seemingly_ mostly functional, but there's a chance I missed something somewhere, and this port also requires extensive testing against FAPI 0.86.0, current QSL, and vanilla for the recently added FAPI Dynamic Registries.

The other changes, including the new Fabric Model Loading API and small additions to the Fabric Transfer API, are seemingly drag-and-drop and did not require significant edits to make functional.

Commit:
Model Loading API v1 (#3145)

This PR deprecates and supersedes the old `fabric-models-v0` module.

## Refactors compared to v0
### Model loading hooks
Mods now register a single `ModelLoadingPlugin` for any change they want to make to the model loading process. (Or multiple plugins if they want to). This replaces the old `ModelLoadingRegistry`.

Here is an example:

```java
ModelLoadingPlugin.register(pluginContext -> {
    // ResourceManager access is provided like in the v0 API, but in a central place, and can be used for shared processing in the plugin.
    ResourceManager manager = pluginContext.resourceManager();

    // Model resource providers still exist, but they are called model resolvers now
    pluginContext.resolveModel().register(context -> {
        // ...
    });
});
```

#### `ModelVariantProvider` -> `BlockStateResolver`
`ModelVariantProvider` was heavily reworked, and is now replaced by the `BlockStateResolver`, with a much better defined contract.

#### `ModelResourceProvider` -> `ModelResolver`
The resource provider is mostly unchanged. The biggest difference is that it is now registered as an event listener. This allows mods to use event phases for ordering between conflicting ~~providers~~ resolvers.

#### Removed custom exception
Additionally, `ModelProviderException` could be thrown by a variant or resource provider in the v0 API. This was not explained in the documentation, and would according to the code stop further processing of the providers and log an error.

In the new API, any `Exception` is caught and logged. If that happens, the other resolvers are still processed. There is no custom `Exception` subclass anymore.

### Helper method to get a `BakedModel` by `Identifier` from the manager
The v0 had such a method in a helper class: `BakedModelManagerHelper#getBakedModel`. It is now interface-injected instead. See `FabricBakedModelManager`.

## New model wrapping hooks
New hooks are added for the various needs of mods that want to override or wrap specific models. Thanks to @embeddedt for contributing an initial version of them!

Here is an example of wrapping the gold model to remove the bottom quads, for example:
```java
ModelLoadingPlugin.register(pluginContext -> {
	// remove bottom face of gold blocks
	pluginContext.modifyModelAfterBake().register(ModelModifier.WRAP_PHASE, (model, context) -> {
		if (context.identifier().getPath().equals("block/gold_block")) {
			return new DownQuadRemovingModel(model);
		} else {
			return model;
		}
	});
});
```

There are 3 events, for the following use cases:
- Wrapping `UnbakedModel`s right when they are loaded. This allows replacing them entirely in dependent models too.
- Wrapping `UnbakedModel`s right before they are baked. This allows replacing them without affecting dependent models (which might not be expecting a model class change).
- Wrapping `BakedModel`s right when they are baked.

Multiple mods have implemented their own version of them. Providing them in Fabric API will make it easier on these mods, and will additionally allow optimization mods that perform on-demand model loading to simply fire the hooks themselves instead of working around injections performed by other mods.

Co-authored-by: embeddedt <42941056+embeddedt@users.noreply.github.com>
Co-authored-by: PepperCode1 <44146161+PepperCode1@users.noreply.github.com>
Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

Juuxel committed:
Add dynamic registry API (#3163)

* Add API for adding custom dynamic registries

Closes #1012, supersedes #1031 and #2719.

* Add missing license headers

* Clarify RegistryLoaderMixin namespace injection

* Replace event with static registration, add skeleton for sorting registries

* Fix typo

* Refactor event phase sorting system for use with dynamic registries (#1)

* Make minor changes to Technici4n's PR

* Add test for nested dynamic objects

* Revert "Add test for nested dynamic objects"

This reverts commit 486e3e1ce097ff66a4018c71d1101d3ee5427d10.

* Revert "Make minor changes to Technici4n's PR"

This reverts commit 741bd52c1e76bf89519d28e6d6b67a6dd04c3fcb.

* Revert "Refactor event phase sorting system for use with dynamic registries (#1)"

This reverts commit bb7c8b8790507bc4095f1963f4e2849d904081b4.

* Remove sorting API

* Add support for defaulted dynamic registries

* Re-add test for nested dynamic objects

* Add missing license headers

* Fix typo

* Remove defaulted dynamic registries; flatten registration methods

* Remove last reference to registry sorting

* Add option to skip syncing for empty dynregs

* Update DynamicRegistrySyncOption docs

Co-authored-by: Technici4n <13494793+Technici4n@users.noreply.github.com>

* Address review feedback

* Add registry namespace to tag paths for modded registries

* Move dynamic registry tests into their own class for readibility

* Finish DynamicRegistries doc

* Only apply tag change to dynamic registries

* Fix checkstyle

* Update fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistries.java

Co-authored-by: Technici4n <13494793+Technici4n@users.noreply.github.com>

---------

Co-authored-by: Technici4n <13494793+Technici4n@users.noreply.github.com>

kevinthegreat1 committed:
Fix incorrect behavior with ClientMessageEvents and CommandApi (#3182)

Commit:
Make event phase ordering logic usable in other contexts (internally only) (#3183)

* Make event phase ordering logic usable in other contexts (internally only)

* Rename and move to toposort package

SquidDev committed:
Fix crash when beehive is broken by fake player (#3190)

* Fix crash when beehive is broken by fake player

When a beehive is broken, every nearby bee targets a random player.
However, if there are no nearby players, the game crashes.

This should not occur under normal (vanilla) conditions. However, if a
beehive is broken by a fake player there are no players in range, and so
we see a crash.

* Checkstyle, my beloved

* Remove public modifier

* See see see

Commit:
Fix Indigo handling of sculk sensor AO (#3200)

* Fix Indigo handling of sculk sensor AO. Fixes #3153

* Split offset and mean formula lighting config options

Shnupbups committed:
Some more TAWs for block creation and block loot tables (#3201)

* Some more TAWs

Added some TAWs for various methods in `Blocks` used to create certain types of blocks that mods may want to also create. Using these methods will allow them to ensure they have all the right block settings they need to stay consistent with vanilla blocks that use the same methods.

* Add some fields from BlockLootTableGenerator

useful fields from BlockLootTableGenerator. also cleaned up the datagen api's accesswidener file

maityyy committed:
Add the cherry grove to the floral biome tag (#3202)

unroman committed:
Ukrainian translations (#3169)

* Add files via upload

* Add files via upload

Commit:
A few transfer API improvements and deprecations (#3204)

* A few transfer API improvements and deprecations

* Forward implementation of deprecated methods

modmuss50 committed:
Bump version

rinOfTheStars committed:
Merge remote-tracking branch 'fabric-remote/1.20.1' into 0.86.0-upstream-compat

# Conflicts:
#	deprecated/fabric-models-v0/src/client/resources/assets/quilted_fabric_models_v0/icon.png
#	deprecated/fabric-models-v0/src/client/resources/fabric.mod.json
#	deprecated/fabric-models-v0/src/client/resources/quilt.mod.json
#	deprecated/fabric-models-v0/src/client/resources/quilted_fabric_models_v0.mixins.json
#	fabric-api-base/src/main/java/net/fabricmc/fabric/impl/base/event/PhaseSorting.java
#	fabric-model-loading-api-v1/build.gradle
#	fabric-model-loading-api-v1/src/client/resources/assets/fabric-model-loading-api-v1/icon.png
#	fabric-models-v0/src/client/java/net/fabricmc/fabric/impl/client/model/ModelLoadingRegistryImpl.java
#	fabric-models-v0/src/client/java/net/fabricmc/fabric/mixin/client/model/ModelLoaderMixin.java
#	fabric-models-v0/src/testmodClient/java/net/fabricmc/fabric/test/model/BakedModelRenderer.java
#	fabric-models-v0/src/testmodClient/java/net/fabricmc/fabric/test/model/ModelTestModClient.java
#	fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/mixin/registry/sync/RegistryLoaderMixin.java
#	fabric-registry-sync-v0/src/main/resources/fabric-registry-sync-v0.accesswidener
#	fabric-registry-sync-v0/src/main/resources/quilted_fabric_registry_sync_v0.mixins.json
#	fabric-renderer-api-v1/src/testmodClient/java/net/fabricmc/fabric/test/renderer/simple/client/FrameModelResourceProvider.java
#	fabric-renderer-api-v1/src/testmodClient/java/net/fabricmc/fabric/test/renderer/simple/client/PillarModelVariantProvider.java
#	fabric-resource-loader-v0/src/main/resources/assets/quilted_fabric_resource_loader_v0/lang/uk_ua.json
#	gradle.properties

rinOfTheStars committed:
Start porting process for real this time
Redid process of setting up QMJ and mixins.json for model_loading_v1, fixed registry sync access wideners
First area of focus is going to be the new fapi dyn. registries, as they will likely require the most substantial work.

rinOfTheStars committed:
Implement compat with Fabric's Dynamic Registry API.

rinOfTheStars committed:
Implement compat with Fabric's Fabric Model Loading API (v1), replacing the old Model API (v0). Basically a drag-and-drop addition.

rinOfTheStars committed:
Remove mixin into QSL's DynamicMetaRegistryImpl that is not actively required, and in fact would potentially cause a breakage with QSL behavior.
- Also began work on smaller compat changes, such as the FakePlayerTests gametest entry for fake player hive breakage.

rinOfTheStars committed:
Fix some QMJ related things

rinOfTheStars committed:
Run checkstyle and header application, apply fix to model-loading-api qmj

eristhea commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/mixin/registry/sync/SerializableRegistriesMixin.java:
```diff
@@ -0,0 +1,55 @@
+	private static Stream<DynamicRegistryManager.Entry<?>> filterNonSyncedEntries(DynamicRegistryManager drm) {
+		return stream(drm).filter(entry -> {
+			//boolean canSkip = DynamicRegistriesImpl.SKIP_EMPTY_SYNC_REGISTRIES.contains(entry.key());
+			//return !canSkip || entry.value().size() > 0;
```
maybe a comment explaining this? the bridge seems very solid though!

rinOfTheStars replied:
I'm honestly not sure if it even works as a proper bridge, mainly because the Registry Sync API is experimental and ~~undocumented~~ sparsely documented—mainly went on a hunch when writing this... 

Instead of pinging patbox I'm going to look at the Sync API impl first, see if I can figure out wtf it's doing, and then ping him if I can't

rinOfTheStars replied:
Does this work as a bridge for FAPI's version of optional Dynamic Registries or will it need a fix on QSL's side to properly support? @Patbox

The full relevant code is at https://github.com/cocona20xx/quilted-fabric-api/blob/0.86.0-upstream-compat/fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/mixin/registry/sync/SerializableRegistriesMixin.java, since commit view is a bit annoying to actually get a look at code in.

I *think* it'll work, if I'm understanding QSL's `RegistrySynchronization#isEntryPresent` correctly, but I think it's best to actually ask you since you're the author of said API and probably know best here.

Patbox replied:
Quilt's registry sync was always intended to be only used for static registries. It also uses flag system for entries. So if entry is optional, isEntryPresent will return whatever true if client has that optional entry, otherwise not

rinOfTheStars replied:
Will be improved in an upcoming commit thanks to QSL changes.

eristhea commented on fabric-registry-sync-v0/src/testmod/java/net/fabricmc/fabric/test/registry/sync/RegistrySyncTest.java:
```diff
@@ -106,9 +106,9 @@ public void onInitialize() {
-		Registry.register(Registries.STATUS_EFFECT, new Identifier("test", "status_effect"),
-				StatusEffectAccessor.createNewStatusEffect(StatusEffectCategory.NEUTRAL, 0xffff9900)
-		);
+		//Registry.register(Registries.STATUS_EFFECT, new Identifier("test", "status_effect"),
```
why?

rinOfTheStars replied:
I'm looking at it now and I have no clue why that happened—my guess is the checkstyle autoapplier did it? I'll revert it in a bit.

rinOfTheStars replied:
Fixed as of latest commit (80ec887)

eristhea reviewed (commented):
thank you so much for this! as far as I can tell the dynamic registry api is unused in mods so can't really be tested, but this port looks good

rinOfTheStars committed:
Revert weird change to RegistrySyncTest (blame codestyle auto-application?? I have no clue tbh)

rinOfTheStars commented:
Waiting on https://github.com/QuiltMC/quilt-standard-libraries/pull/329 and https://github.com/QuiltMC/quilt-standard-libraries/pull/330 to be merged and a new minor release of QSL to be released containing said fixes/additions to continue on this, as both features will heavily simply what QFAPI needs to come in sync with 0.86.0 and 0.86.1

On that subject, should QFAPI skip 0.86.0 and only provide support for 0.86.1? The latter is pretty much all bugfixes, so it seems like a reasonable thing to do, but I'm not sure.

eristhea commented:
yes, we skip versions that release between updates so we'll go straight to `0.86.1`!

rinOfTheStars commented:
While waiting for QSL prs 330 and 329 to be merged, going to look into fixing ~~#112~~, ~~#109~~, ~~#108~~, and possibly #69, since those aren't reliant on QSL changes
edit: 112 apparently is already fixed. guh. 
edit 2 ~ tokyo drift: Found the fix for #109, it's in https://github.com/QuiltMC/quilt-standard-libraries/pull/333
edit the third: #108 doesn't seem to actually be real in pure QSL + QFAPI, hopefully

EnnuiL commented on deprecated/fabric-models-v0/build.gradle:
```diff
@@ -0,0 +1,7 @@
+archivesBaseName = "fabric-models-v0"
+version = getSubprojectVersion(project)
```
```suggestion
archivesBaseName = "quilted_fabric_models_v0"
version = getSubprojectVersion(project)
upstream_version = getSubprojectUpstreamVersion(project)
```

EnnuiL commented on fabric-model-loading-api-v1/src/client/resources/quilt.mod.json:
```diff
@@ -0,0 +1,50 @@
+	},
+	"quilt_loom": {
+		"injected_interfaces": {
+			"net/minecraft/class_1092": [ "net/fabricmc/fabric/api/client/model/loading/v1/FabricBakedModelManager" ]
```
```suggestion
			"net/minecraft/class_1092": ["net/fabricmc/fabric/api/client/model/loading/v1/FabricBakedModelManager"]
```

EnnuiL commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistries.java:
```diff
@@ -0,0 +1,159 @@
+import net.fabricmc.fabric.impl.registry.sync.DynamicRegistriesImpl;
+
+/**
+ *
+ * @deprecated Use Quilt Registry API's {@link org.quiltmc.qsl.registry.api.dynamic.DynamicMetaRegistry} instead.
+ *
+ * <p>Contains methods for registering and accessing dynamic {@linkplain Registry registries}.
```
Actually, instead of doing this (which breaks Fabric's checkstyle check):
```suggestion
 * Contains methods for registering and accessing dynamic {@linkplain Registry registries}.
```

EnnuiL commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistries.java:
```diff
@@ -0,0 +1,159 @@
+ * // Option 2b: Register a synced registry with a different network codec
+ * // @link substring=registerSynced target="#registerSynced(RegistryKey, Codec, Codec, SyncOption...)":
+ * DynamicRegistries.registerSynced(MY_DATA_KEY, MyData.CODEC, MyData.NETWORK_CODEC);
+ * }
+ */
```
I'd do this
```suggestion
 * }
 *
 * @deprecated Use Quilt Registry API's {@link org.quiltmc.qsl.registry.api.dynamic.DynamicMetaRegistry} instead.
 */
```

EnnuiL commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistries.java:
```diff
@@ -0,0 +1,159 @@
+ * DynamicRegistries.registerSynced(MY_DATA_KEY, MyData.CODEC, MyData.NETWORK_CODEC);
+ * }
+ */
+@Deprecated()
```
oh, also,
```suggestion
@Deprecated
```

EnnuiL commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/DynamicRegistriesImpl.java:
```diff
@@ -0,0 +1,79 @@
+
+		if (!DYNAMIC_REGISTRY_KEYS.add(key)) {
+			throw new IllegalArgumentException("Dynamic registry " + key + " has already been registered!");
+		}
```
You did say that you have more work to do; i won't poke more about the checkstyle violations; Don't forget to address them! or else it won't build

EnnuiL commented on fabric-registry-sync-v0/src/main/resources/quilted_fabric_registry_sync_v0.mixins.json:
```diff
@@ -1,19 +1,22 @@
   "required": true,
   "package": "net.fabricmc.fabric.mixin.registry.sync",
-  "compatibilityLevel": "JAVA_17",
+  "compatibilityLevel": "JAVA_16",
```
nonononono
```suggestion
  "compatibilityLevel": "JAVA_17",
```

rinOfTheStars replied:
I have no clue how that got changed what the hell

EnnuiL commented on fabric-registry-sync-v0/src/main/resources/quilted_fabric_registry_sync_v0.mixins.json:
```diff
@@ -1,19 +1,22 @@
-    "StructuresToConfiguredStructuresFixMixin"
+    "StructuresToConfiguredStructuresFixMixin",
+    "TagManagerLoaderMixin",
+    "quilt.DynamicRegistryManagerSetupContextImplMixin",
+    "quilt.workaround.DynamicRegistryManagerSetupContextImplMixin"
```
i don't know how to feel about the reorder; maybe i'll let it pass

EnnuiL commented on gradle.properties:
```diff
@@ -3,8 +3,8 @@ org.gradle.parallel=true
 archivesBaseName=quilted_fabric_api
-version=7.0.6
-upstream_version=0.85.0
+version=7.0.7
```
keep it 7.0.6 for now, it's a PR;
```suggestion
version=7.0.6
```
it's not going to be 7.0.7 anyway

EnnuiL commented on src/main/resources/quilt.mod.json:
```diff
@@ -114,6 +114,10 @@
 				"id": "quilted_fabric_rendering_v0",
 				"optional": true
 			},
+			{
+				"id": "quilted_fabric_models_v0",
+				"optional": true
+			},
```
once again, i wish this was alphabetically sorted; but i can address that post-merge :p

EnnuiL reviewed (changes requested):
Looks like a solid upstream sync to me! (at least ignoring the fact that it has pending changes)
I did see few things that look small but that are actually important to fix, so here they are:

rinOfTheStars committed:
Apply suggestions from code review

Co-authored-by: Ennui Langeweile <85590273+EnnuiL@users.noreply.github.com>

rinOfTheStars committed:
Upgrade to upstream FAPI 0.86.1 & QSL 6.1.0
- Involved manually copy-pasting in the changes from the following FAPI upstream commits:
-- https://github.com/FabricMC/fabric/pull/3222 (one-to-one)
-- https://github.com/FabricMC/fabric/pull/3216 (required some modification in FabricDynamicRegistryProvider to quiltify)
-- https://github.com/FabricMC/fabric/pull/3212 (one-to-one other than converting FMJ additions to QMJ)
-- https://github.com/FabricMC/fabric/pull/3219 does not have to be added as it is not an issue with QSL's 0.86.1 impl of the same concept
- Removed/rewrote parts of the QFAPI impl of FAPI's Dynamic Registry API that are better supported by using QSL's equivalents
- Applied checkstyle and licensing checks. No licensing changes were made, though checkstyle changes were.

rinOfTheStars commented:
With the latest commit, QFAPI should be ready for testing and one final review pass! Going to set this as a ready PR now.

rinOfTheStars committed:
Remove mixins removed in last commit from quilted_fabric_registry_sync_v0.mixins.json

eristhea reviewed (approved):
looks great!

rinOfTheStars committed:
Remove SaveLoadingMixin
